File: python/src/contriwork_config_core/schema.py
"""SchemaAdapter protocol and the built-in Pydantic adapter.

ConfigCore does not bundle a schema library. An adapter is one method wide —
callers who want zod-like validation, attrs, msgspec, etc. can write their
own. The built-in :class:`PydanticAdapter` wraps a pydantic ``BaseModel``
class.
"""

from __future__ import annotations

from typing import Any, Protocol

from pydantic import BaseModel, ValidationError

from .errors import ValidationFailed


class SchemaAdapter[T](Protocol):
    """Validate a dict against a schema and return a typed instance."""

    def validate(self, data: dict[str, Any]) -> T:
        """Validate ``data``; return the typed result or raise :class:`ValidationFailed`."""
        ...


class PydanticAdapter[ModelT: BaseModel]:
    """Adapt a pydantic ``BaseModel`` class to the :class:`SchemaAdapter` protocol."""

    def __init__(self, model: type[ModelT]) -> None:
        self._model = model

    def validate(self, data: dict[str, Any]) -> ModelT:
        try:
            return self._model.model_validate(data)
        except ValidationError as e:
            raise ValidationFailed(
                f"config failed validation against {self._model.__name__}: "
                f"{e.error_count()} error(s)",
                details=e.errors(),
            ) from e

File: typescript/src/schema.ts
/**
 * SchemaAdapter interface and the built-in zod adapter.
 *
 * ConfigCore does not bundle a schema library. An adapter is one method
 * wide — callers who prefer valibot, arktype, ajv, etc. can write their
 * own. The built-in {@link ZodAdapter} wraps a `zod` schema.
 */

import type { z, ZodTypeAny } from "zod";

import { ValidationFailed } from "./errors.js";

/** Validate a dict against a schema and return a typed instance. */
export interface SchemaAdapter<T> {
  validate(data: Record<string, unknown>): T;
}

/** Adapt a zod schema to the {@link SchemaAdapter} interface. */
export class ZodAdapter<S extends ZodTypeAny> implements SchemaAdapter<
  z.infer<S>
> {
  constructor(private readonly schema: S) {}

  validate(data: Record<string, unknown>): z.infer<S> {
    const result = this.schema.safeParse(data);
    if (!result.success) {
      throw new ValidationFailed(
        `config failed validation: ${result.error.issues.length} issue(s)`,
        result.error.issues,
      );
    }
    return result.data;
  }
}
